File: benchmark/template/programs/java/micro/mb07.java
import java.util.Arrays;

/** MB-07 — sorting: bottom-up iterative merge sort, ascending, stable, ping-pong buffers. */
class Main {

    private static final int N = 2_000_000;
    private static final int R = 4;
    private static final int SEED = 20267917;

    /** Park-Miller (Lehmer / MINSTD) generator: state = (state * 48271) mod 2147483647. */
    private static final class Lcg {
        private long state;

        Lcg(long seed) {
            this.state = seed;
        }

        long nextInt() {
            state = (state * 48271L) % 2147483647L;
            return state;
        }
    }

    private static void msort(long[] a, long[] buf, int n) {
        long[] src = a;
        long[] dst = buf;
        int width = 1;
        while (width < n) {
            int lo = 0;
            while (lo < n) {
                int mid = Math.min(lo + width, n);
                int hi = Math.min(lo + 2 * width, n);
                int i = lo;
                int j = mid;
                int k = lo;
                while (i < mid && j < hi) {
                    if (src[i] <= src[j]) {
                        dst[k] = src[i];
                        i++;
                    } else {
                        dst[k] = src[j];
                        j++;
                    }
                    k++;
                }
                while (i < mid) {
                    dst[k] = src[i];
                    i++;
                    k++;
                }
                while (j < hi) {
                    dst[k] = src[j];
                    j++;
                    k++;
                }
                lo += 2 * width;
            }
            long[] tmp = src;
            src = dst;
            dst = tmp;
            width *= 2;
        }
        if (src != a) {
            System.arraycopy(src, 0, a, 0, n);
        }
    }

    /** The reported fields of one whole workload execution. */
    record Result(long total, long ssum, long inv) { }

    /** The whole workload body (methodology 06 section 5.2): re-seeds and regenerates src. */
    static Result workload() {
        Lcg g = new Lcg(SEED);
        long[] src = new long[N];
        for (int i = 0; i < N; i++) {
            src[i] = g.nextInt();
        }
        long[] buf = new long[N];

        long total = 0;
        long ssum = 0;
        long inv = 0;
        for (int r = 0; r < R; r++) {
            src[r] = src[r] + 1;
            long[] a = Arrays.copyOf(src, N);
            msort(a, buf, N);

            long chk = 0;
            for (int i = 0; i < N; i++) {
                chk = (chk * 31 + (a[i] % 1000003)) % 1000003;
            }
            total = (total * 7 + chk) % 1000003;

            ssum = 0;
            for (int i = 0; i < N; i++) {
                ssum += a[i];
            }
            for (int i = 1; i < N; i++) {
                if (a[i - 1] > a[i]) {
                    inv++;
                }
            }
        }

        return new Result(total, ssum, inv);
    }

    public static void main(String[] args) {
        String mode = args.length > 0 ? args[0] : "once";
        int k = args.length > 1 ? Integer.parseInt(args[1]) : 7;

        Result res;
        if (mode.equals("steady")) {
            res = null;
            for (int i = 0; i < k; i++) {
                long t0 = System.nanoTime(); // section 5.2 monotonic clock for Java
                res = workload();
                long elapsed = System.nanoTime() - t0;
                System.out.println("ITER " + i + " " + elapsed);
                System.out.flush(); // section 5.2: flushed immediately
            }
        } else {
            res = workload();
        }

        System.out.println("MB07 " + res.total() + " " + res.ssum() + " " + res.inv());
    }
}

File: benchmark/template/programs/zig/micro/mb07.zig
// MB-07 - Sorting: bottom-up iterative merge sort, ascending, stable, ping-pong buffers.
const std = @import("std");

const n: usize = 2000000;
const rounds: usize = 4;
const seed: i64 = 20267917;

/// Lehmer / MINSTD generator: s = (48271 * s) mod (2^31 - 1).
const Lcg = struct {
    state: i64,

    fn nextInt(self: *Lcg) i64 {
        self.state = @rem(48271 * self.state, 2147483647); // operands non-negative: rem == mod (section 2.1)
        return self.state;
    }
};

fn msort(a: []i64, buf: []i64) void {
    var src = a;
    var dst = buf;

    var width: usize = 1;
    while (width < a.len) : (width *= 2) {
        var lo: usize = 0;
        while (lo < a.len) : (lo += 2 * width) {
            const mid = @min(lo + width, a.len);
            const hi = @min(lo + 2 * width, a.len);
            var i = lo;
            var j = mid;
            var k = lo;
            while (i < mid and j < hi) : (k += 1) {
                if (src[i] <= src[j]) {
                    dst[k] = src[i];
                    i += 1;
                } else {
                    dst[k] = src[j];
                    j += 1;
                }
            }
            while (i < mid) : ({
                i += 1;
                k += 1;
            }) dst[k] = src[i];
            while (j < hi) : ({
                j += 1;
                k += 1;
            }) dst[k] = src[j];
        }
        std.mem.swap([]i64, &src, &dst);
    }

    if (src.ptr != a.ptr) @memcpy(a, src);
}

const Result = struct { total: i64, ssum: i64, inv: i64 };

/// The whole workload body (section 5.2: `steady` re-runs this, so the generator
/// is re-seeded and `src` is regenerated on every iteration).
/// Section 4.12(a) pins zig's 64-bit integer array to `[]i64` from `std.heap.c_allocator`.
fn workload(gpa: std.mem.Allocator) !Result {
    const src = try gpa.alloc(i64, n);
    defer gpa.free(src);
    var g = Lcg{ .state = seed };
    for (src) |*e| e.* = g.nextInt();

    const buf = try gpa.alloc(i64, n);
    defer gpa.free(buf);

    var total: i64 = 0;
    var ssum: i64 = 0;
    var inv: i64 = 0;

    for (0..rounds) |r| {
        src[r] += 1; // anti-elimination
        const a = try gpa.dupe(i64, src); // `a = copy of src`, fresh every round
        defer gpa.free(a);
        msort(a, buf);

        var chk: i64 = 0;
        for (a) |v| chk = @rem(chk * 31 + @rem(v, 1000003), 1000003);
        total = @rem(total * 7 + chk, 1000003);

        ssum = 0;
        for (a) |v| ssum += v;

        for (1..n) |i| {
            if (a[i - 1] > a[i]) inv += 1;
        }
    }
    return .{ .total = total, .ssum = ssum, .inv = inv };
}

pub fn main(init: std.process.Init.Minimal) !void {
    var threaded = std.Io.Threaded.init_single_threaded;
    const io = threaded.io();
    const stdout = std.Io.File.stdout();
    const gpa = std.heap.c_allocator;

    // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
    var mode: []const u8 = "once";
    var k_count: usize = 7;
    var args = init.args.iterate();
    _ = args.next();
    if (args.next()) |a| {
        mode = a;
        if (args.next()) |b| k_count = std.fmt.parseInt(usize, b, 10) catch 7;
    }
    if (k_count < 1) k_count = 7;

    var res: Result = undefined;
    if (std.mem.eql(u8, mode, "steady")) {
        var ibuf: [64]u8 = undefined;
        for (0..k_count) |k| {
            const t0 = std.Io.Clock.awake.now(io);
            res = try workload(gpa);
            const t1 = std.Io.Clock.awake.now(io);
            std.mem.doNotOptimizeAway(res);
            const ns: u64 = @intCast(t0.durationTo(t1).nanoseconds);
            const iter_line = try std.fmt.bufPrint(&ibuf, "ITER {d} {d}\n", .{ k, ns });
            try stdout.writeStreamingAll(io, iter_line);
        }
    } else {
        res = try workload(gpa);
    }

    var out: [128]u8 = undefined;
    const line = try std.fmt.bufPrint(&out, "MB07 {d} {d} {d}\n", .{ res.total, res.ssum, res.inv });
    try stdout.writeStreamingAll(io, line);
}
